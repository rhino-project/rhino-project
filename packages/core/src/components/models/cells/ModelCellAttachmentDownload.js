import { useCallback, useMemo } from 'react';
import { useGlobalComponentForAttribute } from '../../../hooks/overrides';
import { CellLink } from '../../table/cells/CellLink';

export const ModelCellAttachmentDownloadBase = ({
  children,
  getValue,
  ...props
}) => {
  const syntheticGetValue = useCallback(
    () => getValue()?.url_attachment,
    [getValue]
  );
  const linkText = useMemo(
    () => children || getValue()?.display_name,
    [children, getValue]
  );

  return (
    <CellLink getValue={syntheticGetValue} {...props}>
      {linkText}
    </CellLink>
  );
};

export const ModelCellAttachmentDownload = (props) =>
  useGlobalComponentForAttribute(
    'ModelCellAttachmentDownload',
    ModelCellAttachmentDownloadBase,
    props
  );
