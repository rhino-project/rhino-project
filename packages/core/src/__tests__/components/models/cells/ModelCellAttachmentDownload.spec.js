import { render } from '@testing-library/react';
import ModelCellAttachmentDownload from '../../../../components/models/cells/ModelCellAttachmentDownload';
import { sharedCellTests } from './sharedCellTests';

describe('ModelCellAttachmentDownload', () => {
  sharedCellTests(ModelCellAttachmentDownload);

  it(`uses url for link text as fallback`, () => {
    const getValue = () => ({
      url_attachment: 'https://example.com/image.png'
    });

    const { asFragment } = render(
      <ModelCellAttachmentDownload
        model={{
          properties: {
            dummy: {
              type: 'reference'
            }
          }
        }}
        path="dummy"
        getValue={getValue}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it(`uses display_name for link text`, () => {
    const getValue = () => ({
      display_name: 'foo',
      url_attachment: 'https://example.com/image.png'
    });

    const { asFragment } = render(
      <ModelCellAttachmentDownload
        model={{
          properties: {
            dummy: {
              type: 'reference'
            }
          }
        }}
        alt="foo"
        path="dummy"
        getValue={getValue}
      >
        foo
      </ModelCellAttachmentDownload>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
