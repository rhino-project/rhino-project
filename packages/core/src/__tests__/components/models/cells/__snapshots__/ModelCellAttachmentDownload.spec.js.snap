// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModelCellAttachmentDownload > should render the component with className from inherited props 1`] = `
<DocumentFragment>
  <div
    class="dummy-class"
  >
    baz
  </div>
</DocumentFragment>
`;

exports[`ModelCellAttachmentDownload > should render the empty text when value is nullish 1`] = `
<DocumentFragment>
  <div>
    -
  </div>
</DocumentFragment>
`;

exports[`ModelCellAttachmentDownload > should render the overridden empty text when value is nullish 1`] = `
<DocumentFragment>
  <div>
    baz
  </div>
</DocumentFragment>
`;

exports[`ModelCellAttachmentDownload > should render with global override shorthand 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;

exports[`ModelCellAttachmentDownload > should render with global override shorthand for model 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;

exports[`ModelCellAttachmentDownload > should render with global override shorthand for model and attribute 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;

exports[`ModelCellAttachmentDownload > uses display_name for link text 1`] = `
<DocumentFragment>
  <a
    alt="foo"
    href="https://example.com/image.png"
  >
    foo
  </a>
</DocumentFragment>
`;

exports[`ModelCellAttachmentDownload > uses url for link text as fallback 1`] = `
<DocumentFragment>
  <a
    href="https://example.com/image.png"
  >
    https://example.com/image.png
  </a>
</DocumentFragment>
`;
